receiver/zookeeperreceiver/scraper.go

```diff
@@ -16,26 +16,160 @@ package zookeeperreceiver
+	if err != nil {
+		return nil, err
+	}
+
+	if config.Timeout <= 0 {
+		return nil, errors.New("timeout must be a positive duration")
+	}
+
+	return &zookeeperMetricsScraper{
+		logger: logger,
+		config: config,
+	}, nil
 }
 
 func (z *zookeeperMetricsScraper) Initialize(_ context.Context) error {
 	return nil
 }
 
 func (z *zookeeperMetricsScraper) Close(_ context.Context) error {
+	z.cancel()
 	return nil
 }
 
-func (z *zookeeperMetricsScraper) Scrape(_ context.Context) (pdata.ResourceMetricsSlice, error) {
-	return pdata.ResourceMetricsSlice{}, nil
+func (z *zookeeperMetricsScraper) Scrape(ctx context.Context) (pdata.ResourceMetricsSlice, error) {
+	var ctxWithTimeout context.Context
+	ctxWithTimeout, z.cancel = context.WithTimeout(ctx, z.config.Timeout)
+
+	conn, err := z.config.Dial()
+	if err != nil {
+		z.logger.Error("failed to establish connection",
+			zap.String("endpoint", z.config.Endpoint),
+			zap.Error(err),
+		)
+		return pdata.ResourceMetricsSlice{}, err
+	}
+	defer conn.Close()
+
+	deadline, ok := ctxWithTimeout.Deadline()
+	if ok {
+		conn.SetDeadline(deadline)
+	}
+
+	rawStats, serverInfo := z.getRawMNTRStats(conn)
+
+	metricsSlice := pdata.NewMetricSlice()
+	metricsSlice.Resize(len(rawStats))
+
+	now := timeToUnixNano(time.Now())
+
+	var idx int
+	for _, rawStat := range rawStats {
+		metricDescriptor := getOTLPMetricDescriptor(rawStat.metric)
+		initializeMetric(metricsSlice.At(idx), metricDescriptor, now, rawStat.val)
+		idx++
+	}
+
+	return resourceMetricSlice(serverInfo, metricsSlice), nil
+}
+
+func resourceMetricSlice(serverInfo serverInfo, metricsSlice pdata.MetricSlice) pdata.ResourceMetricsSlice {
+	resourceMetricSlice := pdata.NewResourceMetricsSlice()
+
+	// If no stats are collected, return empty value.
+	if metricsSlice.Len() == 0 {
+		return resourceMetricSlice
+	}
+
+	resourceMetricSlice.Resize(1)
+	rm := resourceMetricSlice.At(0)
+	rm.Resource().InitEmpty()
+	rm.Resource().Attributes().Insert(serverStateResourceLabel, pdata.NewAttributeValueString(serverInfo.serverState))
+	rm.Resource().Attributes().Insert(zkVersionResourceLabel, pdata.NewAttributeValueString(serverInfo.version))
+
+	ilms := rm.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	ilm := ilms.At(0)
+	metricsSlice.MoveAndAppendTo(ilm.Metrics())
+
+	return resourceMetricSlice
+}
+
+type rawStat struct {
+	metric string
+	val    int64
+}
+
+type serverInfo struct {
+	serverState string
+	version     string
+}
+
+func (z *zookeeperMetricsScraper) getRawMNTRStats(conn net.Conn) ([]rawStat, serverInfo) {
+	scanner := sendCmd(conn, mntrCommand)
+	rawStats := make([]rawStat, 0, maxMetricsLen)
+	info := serverInfo{}
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		parts := zookeeperFormatRE.FindStringSubmatch(line)
+		if len(parts) != 3 {
+			z.logger.Warn("unexpected line in response",
+				zap.String("command", mntrCommand),
+				zap.String("line", line),
+			)
+			continue
+		}
+
+		if zkVersionKey == parts[1] {
```

Comment:
What do you think about consolidating these string comparisons and the switch statement in `getOTLPMetricDescriptor` into a single map lookup that returns a handler per key? Maybe something like a `map[string]handlerFunc`?

Reply:
Moved this check into a switch statement now instead. I believe it's kind of similar to using a map to do the lookup. Let me know what you think.